File: mcp-code-intelligence-kotlin/src/main/kotlin/com/codeintel/analyzers/entrypoints/detectors/EventDetector.kt
/**
 * KSA-162: Event/MQ/Scheduled Handler Detector.
 */
package com.codeintel.analyzers.entrypoints.detectors

import com.codeintel.analyzers.entrypoints.*

private data class EventIndicator(val pattern: String, val type: EntryType)

private val EVENT_INDICATORS = listOf(
    EventIndicator("@EventHandler", EntryType.EVENT_HANDLER),
    EventIndicator("@Scheduled", EntryType.SCHEDULED),
    EventIndicator("@Cron", EntryType.SCHEDULED),
    EventIndicator("on_event", EntryType.EVENT_HANDLER),
    EventIndicator(".on(", EntryType.EVENT_HANDLER),
    EventIndicator(".subscribe(", EntryType.EVENT_HANDLER),
    EventIndicator("@RabbitListener", EntryType.EVENT_HANDLER),
    EventIndicator("@KafkaListener", EntryType.EVENT_HANDLER),
    EventIndicator("@SqsListener", EntryType.EVENT_HANDLER),
    EventIndicator("setInterval", EntryType.SCHEDULED),
    EventIndicator("cron.schedule", EntryType.SCHEDULED),
)

class EventDetector {

    fun detect(symbols: List<SymbolInput>, source: String): List<EntryPoint> {
        val results = mutableListOf<EntryPoint>()
        for (sym in symbols) {
            val decorators = sym.decorators ?: emptyList()
            val ctx = getContext(source, sym.startLine)
            for (indicator in EVENT_INDICATORS) {
                val found = decorators.any { it.contains(indicator.pattern) } ||
                    ctx.contains(indicator.pattern)
                if (found) {
                    val eventName = extractEventName(decorators, ctx, indicator.pattern)
                    results.add(EntryPoint(
                        symbolId = sym.id, symbolName = sym.name,
                        filePath = sym.filePath, startLine = sym.startLine,
                        entryType = indicator.type,
                        eventName = eventName ?: sym.name,
                        confidence = Confidence.Medium,
                    ))
                    break
                }
            }
        }
        return results
    }

    private fun extractEventName(decorators: List<String>, context: String, pattern: String): String? {
        for (dec in decorators) {
            if (dec.contains(pattern)) {
                val match = Regex("['\"`]([^'\"`]+)['\"`]").find(dec)
                if (match != null) return match.groupValues[1]
            }
        }
        val idx = context.indexOf(pattern)
        if (idx >= 0) {
            val after = context.substring(idx + pattern.length)
            val match = Regex("['\"`]([^'\"`]+)['\"`]").find(after)
            if (match != null) return match.groupValues[1]
        }
        return null
    }

    private fun getContext(source: String, startLine: Int): String {
        val lines = source.split("\n")
        val start = maxOf(0, startLine - 2)
        val end = minOf(lines.size, startLine + 3)
        return lines.subList(start, end).joinToString("\n")
    }
}

File: backend/mcp-code-intelligence-python/src/mcp_code_intel/analyzers/entry_points/detectors/event_detector.py
"""KSA-162: Event/Scheduled Handler Detector."""
from __future__ import annotations
import re
from .. import EntryPoint, EntryType, Confidence, SymbolInput

EVENT_INDICATORS = [
    ("@EventHandler", EntryType.EVENT_HANDLER), ("@Scheduled", EntryType.SCHEDULED),
    ("@Cron", EntryType.SCHEDULED), ("on_event", EntryType.EVENT_HANDLER),
    (".on(", EntryType.EVENT_HANDLER), (".subscribe(", EntryType.EVENT_HANDLER),
    ("@RabbitListener", EntryType.EVENT_HANDLER), ("@KafkaListener", EntryType.EVENT_HANDLER),
    ("@SqsListener", EntryType.EVENT_HANDLER), ("setInterval", EntryType.SCHEDULED),
    ("cron.schedule", EntryType.SCHEDULED),
]


class EventDetector:
    def detect(self, symbols: list[SymbolInput], source: str) -> list[EntryPoint]:
        results = []
        for sym in symbols:
            decs = sym.decorators or []
            ctx = self._get_context(source, sym.start_line)
            for pattern, entry_type in EVENT_INDICATORS:
                if any(pattern in d for d in decs) or pattern in ctx:
                    event_name = self._extract_event_name(decs, ctx, pattern) or sym.name
                    results.append(EntryPoint(
                        symbol_id=sym.id, symbol_name=sym.name,
                        file_path=sym.file_path, start_line=sym.start_line,
                        entry_type=entry_type, event_name=event_name,
                        confidence=Confidence.Medium,
                    ))
                    break
        return results

    def _extract_event_name(self, decs: list[str], ctx: str, pattern: str) -> str | None:
        for d in decs:
            if pattern in d:
                m = re.search(r"""['"`]([^'"`]+)['"`]""", d)
                if m: return m.group(1)
        idx = ctx.find(pattern)
        if idx >= 0:
            m = re.search(r"""['"`]([^'"`]+)['"`]""", ctx[idx + len(pattern):])
            if m: return m.group(1)
        return None

    def _get_context(self, source: str, start_line: int) -> str:
        lines = source.split("\n")
        return "\n".join(lines[max(0, start_line - 2):min(len(lines), start_line + 3)])
